### PR TITLE
Fix for remote__data field.

### PR DESCRIPTION
**Pull request by spficklin**
Issue #423

## Type(s) of Change(s)

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API-specific change (fix or addition to an API function)
- [ ] Updates documentation (inline or markdown files)

## Description

As reported in issue #423 the remote__data field was not properly working. The problem was deeper than the error messages reported in the issue.  This  PR contains a variety of fixes that make the field work.  

## Testing?

This test made a fix to the web services so to test this both the remote and local sites need to be on this branch.  An easy workaround is to just use the local site's web services as if it were a remote site.  Follow these steps to test:
1)  Pull the branch for this PR
2)  Make sure the tripal_ws module is enabled.
3)  Add a new Tripal remote site that really points to the local site.  Do this by going to Tripal > Data Storage > Remote Tripal Sites > Configuration.  For my local test machine I used the URL http://localhost and version 'v0.1'.
4)  Go to the Organism content type:  Structure > Tripal Content Types and click manage fields on the Organism content type and create a new field of type 'Remote Tripal Site'.  Set the Remote Tripal Site to the site you created in step 3;  set the query to "Organism?genus=[taxrank__genus]&species=[taxrank__species]";  set the Field to display to 'abbreviation' or if your organisms don't have an abbreviation set it to 'genus' or 'species'. If you have used the NCBI taxonomy importer then you can use 'division' or any other field added by that importer.  For the test, the Controlled Vocabulary Term at th end doesn't matter. Just pick something that doesn't already belong to a field for the content type.
5)  Make sure the new field shows up on the page by clicking on the 'Manage Display' tab at the top and moving it out of the Disabled section.
6)  Go to any organism page that already exists on the site.  You should see the new field that houses data from the "remote site" (actually your local site just repeated).   If you received no errors along the way and the data is shown then all is good.

## Additional Notes (if any):

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by bradfordcondon — COMMENTED** [review]
edit: oops.  I'm not getting results.  let me doublecheck my setup...

**Comment by bradfordcondon** [comment]
I'm not sure if i am the problem or if the code is.  Field is nulling ( i get "There are no results.")


If I dpm `    $data = $this->makeRemoteRequest($query_str);` in the field load, I get null.


```
Notice: Undefined variable: error in tripal_get_remote_content() (line 259 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal_ws/api/tripal_ws.api.inc).
Notice: Undefined index: label in remote__data_formatter->view() (line 104 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal_ws/includes/TripalFields/remote__data/remote__data_formatter.inc).
Notice: Undefined index: ItemPage in remote__data_formatter->view() (line 105 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal_ws/includes/TripalFields/remote__data/remote__data_formatter.inc).
```


I can visit my web services at http://localhost:8083/web-services/

My remote site config is below.


screen shot 2018-06-19 at 5 14 17 pm


Here is my field config:

* query: `Organism?genus=[taxrank__genus]&species=[taxrank__species]`
* field_to_display: `genus`

Can you tell if my setup is incorrect or if there's a problem with the field?

**Comment by almasaeed2010** [comment]
Hello all,

I got the same errors that @bradfordcondon after clearing cache. However, I do get the expected results, i.e the abbreviation field shows up:
screen shot 2018-06-21 at 9 16 37 am

Thanks!

**Comment by almasaeed2010** [comment]
For completeness, here are the errors I get (slightly different than @bradfordcondon):

```
Notice: Undefined variable: field in local__genome_browser->load() (line 128 of /DevSites/Tripal/sites/all/modules/tripal_manage_analyses/includes/TripalFields/local__genome_browser/local__genome_browser.inc).
Notice: Undefined variable: error in tripal_get_remote_content() (line 259 of /DevSites/Tripal/sites/all/modules/tripal/tripal_ws/api/tripal_ws.api.inc).
```

**Comment by spficklin** [comment]
Thanks @almasaeed2010 and @bradfordcondon.

That first error is because you all probably reviewed one of the other pull requests that had the local__genome_browser field in it.  Then when your branch got switched that field went away and left Drupal scratching it's head.  

The second error is indeed a mistake.  I've just pushed a fix.  Can you try again?

**Comment by almasaeed2010** [comment]
After pulling, I am now getting a fatal error:
```
Error: Cannot use object of type stdClass as array in tripal_get_remote_content() (line 257 of /Sites/Tripal/sites/all/modules/tripal/tripal_ws/api/tripal_ws.api.inc).
```

File: tripal_ws/api/tripal_ws.api.inc
Comment (almasaeed2010) [inline tripal_ws/api/tripal_ws.api.inc]: I think this should be `$data->error`?

Comment (almasaeed2010) [reply tripal_ws/api/tripal_ws.api.inc]: Yep confirmed that changing `$data['error']` to `$data->error` resolves all the issues I've encountered

Comment (spficklin) [reply tripal_ws/api/tripal_ws.api.inc]: Whoops sorry about that.

Comment (spficklin) [reply tripal_ws/api/tripal_ws.api.inc]: Can you commit your fix?

Comment (almasaeed2010) [reply tripal_ws/api/tripal_ws.api.inc]: I don't have commit permissions on the repo so I'd have to create a separate PR but @bradfordcondon does so I'll ask if he can push the change 👍

Comment (bradfordcondon) [reply tripal_ws/api/tripal_ws.api.inc]: I committed the change.

**Review by almasaeed2010 — APPROVED** [review]

**Comment by bradfordcondon** [comment]
i think my field is connected wrong, so if it works for abdullah im happy to call it functional.

that said, i get these warnings:

```
Notice: Undefined index: label in remote__data_formatter->view() (line 103 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal_ws/includes/TripalFields/remote__data/remote__data_formatter.inc).
Notice: Undefined index: ItemPage in remote__data_formatter->view() (line 104 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal_ws/includes/TripalFields/remote__data/remote__data_formatter.inc).
```

**Comment by spficklin** [comment]
Thanks for posting the fix. I forgot you didn't have write access @almasaeed2010 .    @bradfordcondon can you post a screenshot of your remote field setup?

**Comment by bradfordcondon** [comment]
@spficklin 
screen shot 2018-06-22 at 3 57 46 pm
screen shot 2018-06-22 at 3 57 50 pm

**Comment by spficklin** [comment]
@bradfordcondon What content type did you attach that remote field to?

**Comment by bradfordcondon** [comment]
organism

**Comment by laceysanderson** [comment]
I'm also having difficulties verifying remote fields. 

My "remote tripal site" is my production site since my dev is behind a firewall on our server. Once I've setup my field using this config: [field_config](https://user-images.githubusercontent.com/1566301/41828883-eb1e22dc-77f4-11e8-878a-ca5f777be954.png), I see the following on my Species (organism) page:

pageview

I've confirmed that I can access the web services on my production site using the query I configured: http://knowpulse.usask.ca/portal/web-services/content/v0.1/Species?genus=Lens&species=culinaris.

When you click on the "[View content on Portal](http://knowpulse.usask.ca/portal/bio_data/20)" link, you are taken to the corresponding page on my production site. Could the issue be because the label for my organism content type is "Species"? Here is my organism content type: [org_edit](https://user-images.githubusercontent.com/1566301/41829074-f2bb9d8e-77f5-11e8-909b-7bb4a148eac0.png)

**Comment by spficklin** [comment]
I have added some improved error/warning messages when things go wrong.  Can you please retry?   One problem you may have is if the site you are pulling remote data from is not using this same branch.  I had to make a fix to web services that this field requires (i.e. case insensitive bundle names).

**Comment by bradfordcondon** [comment]
Thanks @spficklin 
On first load, the field says "no results found".

after reloading the page, I see the error (and a link to the content anyway?)

screen shot 2018-06-26 at 9 35 18 am

This does suggest to me that my web services aren't configured (or i have a firewall and no idea what i'm doing on that front :D)

**Comment by spficklin** [comment]
Okay, I have improved the field so that it gives better warning messages.  It took some refactoring but the error messages should now tell administrators a bit more about what is going on and provide the web service URL that was used to help debug problems.

@bradfordcondon I think the problem you encountered on this last attempt was your content types are not visible by the anonymous user (which is what the field uses to connect).
@laceysanderson the problem you encountered is that you need a fix to web services that only this PR has so it's not going to work on your public Knowpulse site.   If you want to use your dev site, can you use 'localhost' as the server URL?  Without the fix it cannot use case-insensitivity to find your content type :-/

Can you give it one more attempt and see if it's working better for you?

I also applied a minor fix to webservices. I had a content type named 'Organ' and anytime I tried to pull up the "Organism" type it would get the Organ type instead... oops. That's now fixed in this PR too.

**Comment by bradfordcondon** [comment]
I think i followed up on slack but not here.
You are correct: my permission denied error was due to not setting permissiosn for the content type for anonymous users.  After setting the permission, the field works as expected.

I approve the PR.  I'm going to hold up on merging to make sure @laceysanderson 's issue is resolved.

**Review by bradfordcondon — APPROVED** [review]

**Comment by laceysanderson** [comment]
So, I've found a way to configure my remote field to point at my current dev site. However, I still receive the error referring to "Invalid content type: organism" as shown below:
screen shot 2018-06-27 at 2 30 30 pm


I've confirmed this is due to the `Name` for my Organism Content Type being changed to species. Web services seems to use the `Name` in the URL but the remote field is using the `term name`? In my opinion, the **web-services should be switched to use the term name to facilitate my sharing with other sites regardless of what "name" my biologists prefer**.

Note: I have confirmed that if I switch the name back to "Organism" then my field works, so this is not a configuration issue.

## Steps to Reproduce
1. Set-up a remote field on your organism content type and confirm it works.
2. Edit the content type at Admin > Structure > Tripal Content Types > Organism > Edit and change the `Name: Species`.
3. Clear the cache and then refresh your organism page. Your field which was working will now be showing an error.

**Comment by laceysanderson** [comment]
I've moved my error into a separate issue since I've confirmed the fixes in this PR do work as advertised :-) That way we can get these fixes in for the release.

**Comment by spficklin** [comment]
Muchas gracieas!

**Pull request merged**
